### PR TITLE
Keyboard shortcuts for chat commands: adds user interface and code to...

### PR DESCRIPTION
**Pull request by spillz**
...enable keyboard shortcuts as chat commands

replaces PR #1217

### CODE REVIEW

File: src/game.cpp [hunk 1]
```diff
@@ -2264,6 +2264,56 @@ bool Game::getServerContent(bool *aborted)
 				client->nodedefReceived()) {
 			break;
 		}
+		else
```
Comment (est31) [inline src/game.cpp]: could you [fix the style](http://dev.minetest.net/Code_style_guidelines)?

File: src/guiKeyChangeMenu.cpp [hunk 2]
```diff
@@ -354,6 +514,98 @@ bool GUIKeyChangeMenu::OnEvent(const SEvent& event)
 				case GUI_ID_ABORT_BUTTON: //abort
 					quitMenu();
 					return true;
+				case GUI_ID_KEY_ALIAS_ADD:
+					{
```
Comment (est31) [inline src/guiKeyChangeMenu.cpp]: Please remove the extra block. case behaves like goto labels, you don't need its own block.

File: src/guiKeyChangeMenu.cpp [hunk 3]
```diff
@@ -413,5 +722,8 @@ void GUIKeyChangeMenu::init_keys()
 	this->add_key(GUI_ID_KEY_NOCLIP_BUTTON,    wgettext("Toggle noclip"),    "keymap_noclip");
 	this->add_key(GUI_ID_KEY_RANGE_BUTTON,     wgettext("Range select"),     "keymap_rangeselect");
 	this->add_key(GUI_ID_KEY_DUMP_BUTTON,      wgettext("Print stacks"),     "keymap_print_debug_stacks");
+
+	for (size_t i = 0; i<getCommandKeySettingCount(); ++i)
```
Comment (est31) [inline src/guiKeyChangeMenu.cpp]: Could you change this to `i < getCommandKeySettingCount()`?

File: src/guiKeyChangeMenu.h [hunk 4]
```diff
@@ -66,13 +74,29 @@ class GUIKeyChangeMenu: public GUIModalMenu
 
 	void add_key(int id, const wchar_t *button_name, const std::string &setting_name);
 
+	std::wstring keyUsedBy(int id, const KeyPress &key, bool modifier_shift, bool modifier_control);
+	void add_command_alias_key(const KeyCommand &key);
+	void commandComboChanged();
+
```
Comment (est31) [inline src/guiKeyChangeMenu.h]: I'm not sure, is there a space here?

File: src/keycode.cpp [hunk 5]
```diff
@@ -363,3 +365,66 @@ void clearKeyCache()
...
+				k.key = g_settings->get(*it).c_str();
+				if (g_settings->exists("keymap_command_" + aname))
+					k.command = g_settings->get("keymap_command_" + aname).c_str();
+				k.modifier_control = false;
+				k.modifier_shift = false;
+				std::string modifier_list;
+				if (g_settings->exists("keymap_modifiers_" + aname))
+					modifier_list = g_settings->get("keymap_modifiers_" + aname).c_str();
+				size_t pos;
+				while(modifier_list != "")
```
Comment (est31) [inline src/keycode.cpp]: Please change this to `while (modifier_list != "")`

File: src/game.cpp [hunk 6]
```diff
@@ -2264,6 +2264,56 @@ bool Game::getServerContent(bool *aborted)
...
+		{
```
Comment (Zeno-) [inline src/game.cpp]: What's all this do?

Comment (est31) [inline src/game.cpp]: All this should probably moved into the `Game::processKeyboardInput` method.

**Comment by spillz** [comment]
The more important question is: does the code actually work? If it doesn't and you see how to fix it then do so. If it's not working, explain.

Whatever style violations are in changeset are consistent with what was in the original files when the changes have were made. Beyond style, it's quite possible things have gotten screwed up because of incompatible changes to the source files, but it seemed to work OK for me. Instead of pointing out all of these style violation, why don't you just fix it? As someone who isn't a project dev, I have zero interest in making pedantic fixes that others can much more easily make. Either take the code and fix it or reject the code, I frankly no longer care. I think I've more than done my part resubmitting a PR that should have been accepted a long time ago.

File: src/game.cpp (hunk 6 above)
Comment (spillz) [inline src/game.cpp]: Re "What's all this (below) do"

It's supposed to convert a keystroke into a chat command. Looking at the code, I'm not sure that it's in the right location. Perhaps the patch applied badly... Will look tomorrow.

**Comment by Zeno-** [comment]
As a dev I have zero interest in fixing code that doesn't work (I haven't compiled or tested it but I can see that it doesn't). If there are just one or two formatting style issues I'll probably fix them as I merge, but that's not really a requirement.

As for style, the idea behind being pedantic is not because we like to be annoying -- it's because there is currently a "fix style as you go" guideline because a lot of the source code needs style fixes.

> Re "What's all this (below) do"

I know what it's supposed to do but I have some small issues. a) Just glancing at it what the code does, or is intended to do, is not clear (even a simple comment like '// convert a keystroke into a chat command' would have probably been adequate; b) It's in the wrong place and won't work (as already intimated at); c) The "style" (real style, not formatting style) is difficult to read. Basically items (a) and (c) are required so that code is maintainable.

**Comment by spillz** [comment]
Mutually compatible easy solution then!

**Comment by Zeno-** [comment]
I'll fix (rebase) this if people want the feature

**Comment by ObaniGemini** [comment]
Personally, I heavily want this.
This would also enhance PvP system alot and permit to add PvP commands that could be binded.
I over-support this feature that seems so forgotten !

**Comment by celeron55** [comment]
I tested rebasing this. This still fully works. https://github.com/celeron55/minetest/commits/keyboard-shortcuts

**Comment by ObaniGemini** [comment]
oh my god please make a new PR before this gets again forgotten @celeron55 :gem:

**Comment by rubenwardy** [comment]
:+1:

**Comment by celeron55** [comment]
I think the settings for this should be moved out from the "change keys" dialog to a separate dialog which is opened by a single button on the "change keys" dialog. We need the remaining space for other keys; see for example #3846.

**Comment by celeron55** [comment]
Moving to #3856.

**Pull request closed without merging**